lms/static/scripts/frontend_apps/components/BasicLtiLaunchApp.js

```diff
@@ -23,105 +25,155 @@ import Spinner from './Spinner';
+  // and errorState is falsely
+  const showSpinner = () => {
+    return !errorState && fetchCount > 0;
+  };
+
+  // Increment the fetch counter by 1 and clear any
+  // previous error state.
+  const setFetching = () => {
+    count.current += 1;
+    setFetchCount(count.current);
+    setErrorState(null);
+  };
+
+  // Decrement the fetch counter by 1.
+  const setFetched = () => {
+    count.current -= 1;
+    setFetchCount(count.current);
+  };
+
+  /**
+   * Helper to handle thrown errors from from API requests.
+   *
+   * @param {Error} e - Error object from request.
+   * @param {string} errorState - One of "error-fetch",
+   *  "error-authorizing", or "error-report-submission"
+   * @param {boolean} [retry=true] - Can the request be retried?
+   */
+  const handleError = (e, errorState, retry = true) => {
+    setFetched();
+    if (e instanceof ApiError && !e.errorMessage && retry) {
+      setErrorState('error-authorizing');
+    } else {
+      setError(e);
+      setErrorState(errorState);
+    }
+  };
+
+  /**
+   * Fetch the groups from the sync endpoint if `sync` object exists.
+   */
+  const fetchGroups = async () => {
+    if (apiSync) {
+      try {
+        setFetching();
+        const groups = await apiCall({
+          authToken,
+          path: apiSync.path,
+          data: apiSync.data,
+        });
+        rpcServer.resolveGroupFetch(groups);
+        setFetched();
+      } catch (e) {
+        handleError(e, 'error-fetch');
+      }
+    }
+  };
 
   /**
-   * Fetch the URL of the content to display in the iframe.
+   * Fetch the URL of the content to display in the iframe if `viaCallbackUrl`
+   * exists.
    *
    * This will typically be a PDF URL proxied through Via.
    */
-  const fetchContentUrl = useCallback(async () => {
+  const fetchContentUrl = async () => {
     if (!viaCallbackUrl) {
       // If no "callback" URL was supplied for the frontend to use to fetch
       // the URL, then the backend must have provided the Via URL in the
       // initial request, which we'll just use directly.
       return;
     }
-
     try {
-      setLtiLaunchState({
-        ...INITIAL_LTI_LAUNCH_STATE,
-        state: 'fetching-url',
-      });
+      setFetching();
       const { via_url: contentUrl } = await apiCall({
         authToken,
         path: viaCallbackUrl,
       });
-      setLtiLaunchState({
-        ...INITIAL_LTI_LAUNCH_STATE,
-        state: 'fetched-url',
-        contentUrl,
-      });
+      setFetched();
+      setContentUrl(contentUrl);
     } catch (e) {
-      if (e instanceof ApiError && !e.errorMessage) {
-        setLtiLaunchState({
-          ...INITIAL_LTI_LAUNCH_STATE,
-          state: 'authorizing',
-        });
-      } else {
-        setLtiLaunchState({
-          ...INITIAL_LTI_LAUNCH_STATE,
-          state: 'error',
-          error: e,
-          failedAction: 'fetch-url',
-        });
-      }
+      handleError(e, 'error-fetch');
```

Comment:
This `try ... catch` block has the same issue as above in that the `fetchCount` value will not be decremented if an exception occurs.

Reply:
Awe indeed, it will skip the setFetched(); I will use finally.